mountpoint-s3-crt/src/auth/imds_client.rs

```diff
@@ -0,0 +1,119 @@
+    }
+
+    /// Make a get instance type request to ec2 hosts using [ImdsClient].
+    pub fn get_instance_type<F>(&self, callback: F) -> Result<i32, Error>
+    where
+        F: FnOnce(Result<String, Error>) + 'static,
+    {
+        let callback_closure = Box::new(callback);
+        let callback_wrapper = Box::new(ImdsClientGetResourceCallback(Box::new(callback_closure)));
+        let callback_raw_ptr = Box::into_raw(callback_wrapper) as *mut libc::c_void;
+        // must explicit declare the type as function pointer instead of function item if not .
+        let get_resource_callback_fn_ptr: Option<unsafe extern "C" fn(*const aws_byte_buf, i32, *mut libc::c_void)> =
+            Some(imds_client_get_resource_callback);
+
+        // Safety: move the reference where the data store in memory out of box potential leaks memory,
+        // however, the binding callback function will drop it once the corresponding crt function finished the execution.
+        // See `extern "C" fn imds_client_get_resource_callback`.
+        unsafe {
+            let result =
+                aws_imds_client_get_instance_type(self.inner.as_ptr(), get_resource_callback_fn_ptr, callback_raw_ptr);
+            Ok(result)
+        }
+    }
+}
+
+impl Drop for ImdsClient {
+    fn drop(&mut self) {
+        // SAFETY: `self.inner` points to a valid aws_imds_client.
+        unsafe { aws_imds_client_release(self.inner.as_ptr()) };
+    }
+}
+
+impl ImdsClientConfig {
+    /// Create a new [ImdsClientConfig] with default options.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Client bootstrap used for common staples such as event loop group, host resolver, etc.
+    pub fn client_bootstrap(&mut self, client_bootstrap: ClientBootstrap) -> &mut Self {
+        self.inner.bootstrap = client_bootstrap.inner.as_ptr();
+        self.client_bootstrap = Some(client_bootstrap);
+        self
+    }
+
+    /// Retry strategy used to reschedule failed requests
+    pub fn retry_strategy(&mut self, retry_strategy: RetryStrategy) -> &mut Self {
+        self.inner.retry_strategy = retry_strategy.inner.as_ptr();
+        self.retry_strategy = Some(retry_strategy);
+        self
+    }
+}
+
+/// Callback trait for when finished getting ec2 instance type.
+type OnGetResource = Box<dyn FnOnce(Result<String, Error>)>;
+struct ImdsClientGetResourceCallback(OnGetResource);
+
+/// Rust binding for CRT's callback function `aws_imds_client_on_get_resource_callback_fn`.
+unsafe extern "C" fn imds_client_get_resource_callback(
+    resource: *const aws_byte_buf,
+    error_code: i32,
+    user_data: *mut libc::c_void,
+) {
+    let callback = Box::from_raw(user_data as *mut ImdsClientGetResourceCallback).0;
+
+    callback(if 0 != error_code {
+        Err(Error::from(error_code))
+    } else {
+        let resource = std::slice::from_raw_parts((*resource).buffer, (*resource).len).to_vec();
+        Ok(String::from_utf8(resource).unwrap())
+    })
```

Comment:
I wouldn't combine the callback with the return of the if statement, it is making it difficult to follow.

Also, I'd use `.expect(&str)` instead of `.unwrap()` so that we get a meaningful panic message if we do find ourselves in the situation - which hopefully we wouldn't :)

```suggestion
    let result = if 0 != error_code {
        Err(Error::from(error_code))
    } else {
        let resource = std::slice::from_raw_parts((*resource).buffer, (*resource).len).to_vec();
        let resource = String::from_utf8(resource).expect("resource response should be utf8");
        Ok(resource)
    }
    
    callback(result)
```

Reply:
done